src/qdisp/QueryRequest.cc

```diff
@@ -407,74 +448,103 @@ void QueryRequest::ProcessResponseData(XrdSsiErrInfo const& eInfo,
 }
 
 
-void QueryRequest::_processData(JobQuery::Ptr const& jq, int blen, bool last) {
+void QueryRequest::_processData(JobQuery::Ptr const& jq, int blen, bool xrdLast) {
     // It's possible jq and _jobQuery differ, so need to use jq.
-    auto executive = jq->getExecutive();
-    if (executive == nullptr || executive->getCancelled() || executive->isLimitRowComplete()) {
-        if (executive == nullptr || executive->getCancelled()) {
-            LOGS(_log, LOG_LVL_WARN, "QueryRequest::_processData job was cancelled.");
-        } else {
-            int dataIgnored = (executive->incrDataIgnoredCount());
-            if ((dataIgnored - 1)%1000 == 0) {
-                LOGS(_log, LOG_LVL_INFO, "QueryRequest::_processData ignoring, enough rows already "
-                                      << "dataIgnored=" << dataIgnored);
-            }
-        }
+    if (jq->isQueryCancelled()) {
+        LOGS(_log, LOG_LVL_WARN, "QueryRequest::_processData job was cancelled.");
         _errorFinish(true);
         return;
     }
 
     // Get a copy of the shared buffer pointer so askForResponseDataCmd can be deleted.
     ResponseHandler::BufPtr bufPtr = _askForResponseDataCmd->getBufPtr();
     _askForResponseDataCmd.reset(); // No longer need it, and don't want the destructor calling _errorFinish().
+
+    int const protoHeaderSize = proto::ProtoHeaderWrap::getProtoHeaderSize();
+    ResponseHandler::BufPtr nextHeaderBufPtr;
+
+    // Values for these variables to be filled in by flush() calls.
     bool largeResult = false;
     int nextBufSize = 0;
     int resultRows = 0;
-    bool flushOk = jq->getDescription()->respHandler()->flush(blen, bufPtr, last,
-                                                              largeResult, nextBufSize, resultRows);
+    bool last = false;
+    bool flushOk = false;
+
+    // The buffer has 2 parts.
+    // - The first (bytes = blen - ProtoHeaderWrap::getProtheaderSize())
+    //   is the result associated with the previously received header.
+    // - The second is the header for the next message.
+    int respSize = blen - protoHeaderSize;
+    nextHeaderBufPtr = make_shared<vector<char>>(bufPtr->begin() + respSize, bufPtr->end());
```

Comment:
The header is 256 bytes or so, the rest of the buffer is potentially several megabytes. I chose to copy the small part so I could delete the large part as soon as it wasn't needed. I think it also resulted in simpler code.

Reply:
Reducing number of allocations is a good idea in general, and allocating buffers of very different sizes from the same memory arena can cause memory fragmentation.

Reply:
It can be a pretty long time (possibly many minutes) before the need for the nextHeaderBuffer goes away. Considering that there can be thousands of these, I think freeing the large block of memory as early as possible is more important than possible memory fragmentation.